diff --git a/libbb/dump.c b/libbb/dump.c
index fcdee83..c19d5d6 100644
--- a/libbb/dump.c
+++ b/libbb/dump.c
@@ -626,7 +626,7 @@ static NOINLINE void display(priv_dumper_t* dumper)
 							printf(pr->fmt, (char *) bp);
 							break;
 						case F_TEXT:
-							printf(pr->fmt);
+							fputs_stdout(pr->fmt);
 							break;
 						case F_U:
 							conv_u(pr, bp);
@@ -678,7 +678,7 @@ static NOINLINE void display(priv_dumper_t* dumper)
 				printf(pr->fmt, (unsigned long long) dumper->eaddress + dumper->pub.xxd_displayoff);
 				break;
 			case F_TEXT:
-				printf(pr->fmt);
+				fputs_stdout(pr->fmt);
 				break;
 			}
 		}
